[develop] Speed up pimoroni driver.

daskygit opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

This should function almost the same as the old driver. I've avoided the costly floating point calculation in the main loop and implemented throttling. The major difference being the x, y directions are swapped vs the original. (I was following the python library). This can be corrected with a couple of defines.

I don't have prior experience with the trackball on the old driver so some feedback on how it compares would be great. The throttling can be adjusted using the `TRACKBALL_INTERVAL_MS` define, it defaults to 8 (125hz). So likely worth playing with lower values on faster usb polling rates.

The mouse control feels pretty similar to the old driver but with the speed increase being so large it may outweigh the slight difference in feel. I've included a little log of the scan rate while spinning the trackball quickly.

Old driver scan rate: 
```QMK:Onekey Pro Micro:1: matrix scan frequency: 2277
QMK:Onekey Pro Micro:1: matrix scan frequency: 2278
QMK:Onekey Pro Micro:1: matrix scan frequency: 1885
QMK:Onekey Pro Micro:1: matrix scan frequency: 379
QMK:Onekey Pro Micro:1: matrix scan frequency: 682
QMK:Onekey Pro Micro:1: matrix scan frequency: 176
QMK:Onekey Pro Micro:1: matrix scan frequency: 909
QMK:Onekey Pro Micro:1: matrix scan frequency: 158
QMK:Onekey Pro Micro:1: matrix scan frequency: 145
QMK:Onekey Pro Micro:1: matrix scan frequency: 1150
QMK:Onekey Pro Micro:1: matrix scan frequency: 146
QMK:Onekey Pro Micro:1: matrix scan frequency: 486
QMK:Onekey Pro Micro:1: matrix scan frequency: 1698
QMK:Onekey Pro Micro:1: matrix scan frequency: 1690
QMK:Onekey Pro Micro:1: matrix scan frequency: 1370
```

New driver (TRACKBALL_INTERVAL_MS 8) scan rate:
```
QMK:Onekey Pro Micro:1: matrix scan frequency: 9917
QMK:Onekey Pro Micro:1: matrix scan frequency: 9917
QMK:Onekey Pro Micro:1: matrix scan frequency: 9914
QMK:Onekey Pro Micro:1: matrix scan frequency: 9911
QMK:Onekey Pro Micro:1: matrix scan frequency: 9907
QMK:Onekey Pro Micro:1: matrix scan frequency: 9904
QMK:Onekey Pro Micro:1: matrix scan frequency: 9903
QMK:Onekey Pro Micro:1: matrix scan frequency: 9907
QMK:Onekey Pro Micro:1: matrix scan frequency: 9929
QMK:Onekey Pro Micro:1: matrix scan frequency: 9928
QMK:Onekey Pro Micro:1: matrix scan frequency: 9928
QMK:Onekey Pro Micro:1: matrix scan frequency: 9925
QMK:Onekey Pro Micro:1: matrix scan frequency: 9929
```

New driver (TRACKBALL_INTERVAL_MS 1) scan rate:
```
QMK:Onekey Pro Micro:1: matrix scan frequency: 8718
QMK:Onekey Pro Micro:1: matrix scan frequency: 8715
QMK:Onekey Pro Micro:1: matrix scan frequency: 7594
QMK:Onekey Pro Micro:1: matrix scan frequency: 4741
QMK:Onekey Pro Micro:1: matrix scan frequency: 5420
QMK:Onekey Pro Micro:1: matrix scan frequency: 5134
QMK:Onekey Pro Micro:1: matrix scan frequency: 4672
QMK:Onekey Pro Micro:1: matrix scan frequency: 5234
QMK:Onekey Pro Micro:1: matrix scan frequency: 7436
QMK:Onekey Pro Micro:1: matrix scan frequency: 6029
QMK:Onekey Pro Micro:1: matrix scan frequency: 7984
QMK:Onekey Pro Micro:1: matrix scan frequency: 7974
QMK:Onekey Pro Micro:1: matrix scan frequency: 8045
```

I've tested on a promicro and bluepill, the performance benefit exists for both. I wonder if some basic documentation is in order explaining some of the defines with an example on how to implement it.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

zvecr commented on drivers/sensors/pimoroni_trackball.h:
```diff
@@ -22,8 +22,21 @@
 #ifndef TRACKBALL_ADDRESS
 #    define TRACKBALL_ADDRESS 0x0A
 #endif
-#define TRACKBALL_WRITE ((TRACKBALL_ADDRESS << 1) | I2C_WRITE)
-#define TRACKBALL_READ (TRACKBALL_ADDRESS << 1)
+
+#ifndef TRACKBALL_INTERVAL_MS
+#    define TRACKBALL_INTERVAL_MS 8
+#endif
+
+#ifndef TRACKBALL_DEBOUNCE_ONCLICK
+#    define TRACKBALL_DEBOUNCE_ONCLICK 5
+#endif
+typedef struct pimoroni_data {
+    uint8_t left;
+    uint8_t right;
+    uint8_t up;
+    uint8_t down;
+    uint8_t click;
+} pimoroni_data;
```
From a quick scan, I dont see intent in the API to make `TRACKBALL_ADDRESS`, `TRACKBALL_INTERVAL_MS`, `TRACKBALL_DEBOUNCE_ONCLICK` or `pimoroni_data` public. It should be safe to move them to the top of the `.c` file.

daskygit replied:
Thank, noticed `trackball_register_button` isn't actually a thing in the old driver either, so removed it.

daskygit commented:
Much happier with the behaviour compared to the original now. I've added a callback so you can override pointing_device_task, added some extra defines for finer configuration and fixed a couple of logic errors.

I renamed the defines that users may want to configure to have `PIMORONI_` in front. I've also got an example onekey keymap, is it worth including it with this PR?

daskygit commented:
I've added a disconnect count so the code can be flashed to either side of a split without impacting the scan rate if it doesn't exist.

daskygit commented:
Thanks
